File: libraries/Zig/trainer/src/harness.zig
// simulates the same kind of setup that the host would do if running compiled WebAssembly

const std = @import("std");
const client = @import("wasmbot_client");

const stdout = std.io.getStdOut().writer();
const stderr = std.io.getStdErr().writer();
var rng = std.rand.DefaultPrng.init(0);

export fn shutdown() void {
    unreachable; // will just cause a panic, which is fine
}

export fn logFunction(logLevel: i32, msgPtr: usize, msgLen: usize) void {
    const str = liftString(msgPtr, msgLen);
    if (logLevel == 0) {
        stderr.writeAll(str) catch @panic("Couldn't write to console?!");
    } else {
        stdout.writeAll(str) catch @panic("Couldn't write to console?!");
    }
}

export fn getRandomInt(min: i32, max: i32) i32 {
    return rng.random().intRangeLessThan(i32, min, max);
}

pub fn liftString(offset: usize, len: usize) []const u8 {
    const ptr: [*]const u8 = @ptrFromInt(offset);
    var read_len: usize = 0;
    while (read_len < len and ptr[read_len] != 0) {
        read_len += 1;
    }
    return ptr[0..read_len];
}

extern fn clientInitialize() void;
extern fn setup(usize) usize;
extern fn receiveGameParams(usize) bool;
extern fn tick(usize) void;

var reserve_block: []u8 = undefined;

const ValueBlockReturn = struct {
    success: bool,
    mem: []const u8,
};

pub fn simulateSetup(reserve_request: usize) []u8 {
    clientInitialize();
    const reserveOffset = setup(reserve_request);
    if (reserveOffset == 0) {
        return &[0]u8{};
    }
    const reservePtr: [*]u8 = @ptrFromInt(reserveOffset);
    reserve_block = reservePtr[0..reserve_request];
    return reserve_block;
}

pub fn simulateReceiveGameParams(incoming_block: []const u8, offset: usize) ValueBlockReturn {
    @memcpy(reserve_block, incoming_block);

    const bot_ready = receiveGameParams(offset);

    return .{
        .success = bot_ready,
        .mem = reserve_block,
    };
}

pub fn simulateTick(incoming_block: []const u8, offset: usize) []u8 {
    @memcpy(reserve_block, incoming_block);

    tick(offset);

    return reserve_block;
}

File: libraries/C/trainer/harness.c
#include "harness.h"

#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <stdio.h>
#include <stdbool.h>

extern void clientInitialize(void);
extern size_t setup(size_t);
extern bool receiveGameParams(size_t);
extern void tick(size_t);

static uint8_t* _hostReservePtr = NULL;
static size_t _hostReserveLen = 0;

// (seeded in the simulateSetup)
// yes, there are many better RNGs, but this will do for our purposes
int32_t getRandomInt(int32_t min, int32_t max) {
	return rand() % (max - min) + min;
}

void logFunction(int32_t logLevel, size_t msgPtr, size_t msgLen) {
	const char* msg = (const char*)msgPtr;
	if (msg[msgLen] != '\0') {
		// not null-terminated
		char* buffer = (char*)malloc(msgLen + 1);
		if (!buffer) {
			fprintf(stderr, "Could not allocate log memory\n");
			return;
		}
		memcpy(buffer, msg, msgLen);
		buffer[msgLen] = '\0';
		if (logLevel == 0) {
			fprintf(stderr, "%s\n", buffer);
		} else {
			printf("%s\n", buffer);
		}
		free(buffer);
	}
	else {
		if (logLevel == 0) {
			fprintf(stderr, "%s\n", msg);
		}
		else {
			printf("%s\n", msg);
		}
	}
}

void harness_log(const char* msg) {
	logFunction(2, (size_t)msg, strlen(msg));
}

void harness_logErr(const char* msg) {
	logFunction(0, (size_t)msg, strlen(msg));
}

uint8_t* simulateSetup(size_t reserveRequest) {
	srand(time(NULL));

	const size_t reserveOffset = setup(reserveRequest);
	if (reserveOffset == 0) {
		harness_logErr("Could not allocate reserve memory");
		return NULL;
	}

	_hostReservePtr = (uint8_t*)reserveOffset;
	_hostReserveLen = reserveRequest;

	return _hostReservePtr;
}

uint8_t* simulateReceiveGameParams(uint8_t* incomingBlock, size_t offset) {
	memcpy(_hostReservePtr, incomingBlock, _hostReserveLen);

	bool botReady = receiveGameParams(offset);

	if (!botReady) {
		harness_logErr("Bot declined game parameters");
		return NULL;
	}

	return _hostReservePtr;
}

uint8_t* simulateTick(uint8_t* incomingBlock, size_t offset) {
	memcpy(_hostReservePtr, incomingBlock, _hostReserveLen);

	tick(offset);

	return _hostReservePtr;
}
